File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-WATR.Water Type.swift
//
//  WATRRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class WATRRecord: Record {
    public class DATAField {
        public var windVelocity: Float = 0
        public var windDirection: Float = 0
        public var waveAmplitude: Float = 0
        public var waveFrequency: Float = 0
        public var sunPower: Float = 0
        public var reflectivityAmount: Float = 0
        public var fresnelAmount: Float = 0
        public var scrollXSpeed: Float = 0
        public var scrollYSpeed: Float = 0
        public var fogDistance_nearPlane: Float = 0
        public var fogDistance_farPlane: Float = 0
        public var shallowColor: ColorRef4 = ColorRef4_empty
        public var deepColor: ColorRef4 = ColorRef4_empty
        public var reflectionColor: ColorRef4 = ColorRef4_empty
        public var textureBlend: UInt8 = 0
        public var rainSimulator_force: Float = 0
        public var rainSimulator_velocity: Float = 0
        public var rainSimulator_falloff: Float = 0
        public var rainSimulator_dampner: Float = 0
        public var rainSimulator_startingSize: Float = 0
        public var displacementSimulator_force: Float = 0
        public var displacementSimulator_velocity: Float = 0
        public var displacementSimulator_falloff: Float = 0
        public var displacementSimulator_dampner: Float = 0
        public var displacementSimulator_startingSize: Float = 0
        public var damage: UInt16 = 0

        init(_ r: BinaryReader, _ dataSize: Int) {
            guard dataSize != 2 else {
                damage = r.readLEUInt16()
                return
            }
            windVelocity = r.readLESingle()
            windDirection = r.readLESingle()
            waveAmplitude = r.readLESingle()
            waveFrequency = r.readLESingle()
            sunPower = r.readLESingle()
            reflectivityAmount = r.readLESingle()
            fresnelAmount = r.readLESingle()
            scrollXSpeed = r.readLESingle()
            scrollYSpeed = r.readLESingle()
            fogDistance_nearPlane = r.readLESingle()
            guard dataSize != 42 else {
                damage = r.readLEUInt16()
                return
            }
            fogDistance_farPlane = r.readLESingle()
            shallowColor = r.readT(dataSize)
            deepColor = r.readT(dataSize)
            reflectionColor = r.readT(dataSize)
            textureBlend = r.readByte()
            r.skipBytes(3) // Unused
            guard dataSize != 62 else {
                damage = r.readLEUInt16()
                return
            }
            rainSimulator_force = r.readLESingle()
            rainSimulator_velocity = r.readLESingle()
            rainSimulator_falloff = r.readLESingle()
            rainSimulator_dampner = r.readLESingle()
            rainSimulator_startingSize = r.readLESingle()
            displacementSimulator_force = r.readLESingle()
            guard dataSize != 86 else {
                damage = r.readLEUInt16()
                return
            }
            displacementSimulator_velocity = r.readLESingle()
            displacementSimulator_falloff = r.readLESingle()
            displacementSimulator_dampner = r.readLESingle()
            displacementSimulator_startingSize = r.readLESingle()
            damage = r.readLEUInt16()
        }
    }

    public struct GNAMField {
        public let daytime: FormId<WATRRecord> 
        public let nighttime: FormId<WATRRecord> 
        public let underwater: FormId<WATRRecord> 

        init(_ r: BinaryReader, _ dataSize: Int) {
            daytime = FormId<WATRRecord>(r.readLEUInt32())
            nighttime = FormId<WATRRecord>(r.readLEUInt32())
            underwater = FormId<WATRRecord>(r.readLEUInt32())
        }
    }

    public override var description: String { return "WATR: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var TNAM: STRVField! // Texture
    public var ANAM: BYTEField! // Opacity
    public var FNAM: BYTEField! // Flags
    public var MNAM: STRVField! // Material ID
    public var SNAM: FMIDField<SOUNRecord>! // Sound
    public var DATA: DATAField! // DATA
    public var GNAM: GNAMField! // GNAM

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
            case "EDID": EDID = r.readSTRV(dataSize)
            case "TNAM": TNAM = r.readSTRV(dataSize)
            case "ANAM": ANAM = r.readT(dataSize)
            case "FNAM": FNAM = r.readT(dataSize)
            case "MNAM": MNAM = r.readSTRV(dataSize)
            case "SNAM": SNAM = FMIDField<SOUNRecord>(r, dataSize)
            case "DATA": DATA = DATAField(r, dataSize)
            case "GNAM": GNAM = GNAMField(r, dataSize)
            default: return false
        }
        return true
    }
}

File: src/ObjectManager/Object.Tes/FilePacks/Records/045-WATR.Water Type.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class WATRRecord : Record
    {
        public class DATAField
        {
            public float WindVelocity;
            public float WindDirection;
            public float WaveAmplitude;
            public float WaveFrequency;
            public float SunPower;
            public float ReflectivityAmount;
            public float FresnelAmount;
            public float ScrollXSpeed;
            public float ScrollYSpeed;
            public float FogDistance_NearPlane;
            public float FogDistance_FarPlane;
            public ColorRef4 ShallowColor;
            public ColorRef4 DeepColor;
            public ColorRef4 ReflectionColor;
            public byte TextureBlend;
            public float RainSimulator_Force;
            public float RainSimulator_Velocity;
            public float RainSimulator_Falloff;
            public float RainSimulator_Dampner;
            public float RainSimulator_StartingSize;
            public float DisplacementSimulator_Force;
            public float DisplacementSimulator_Velocity;
            public float DisplacementSimulator_Falloff;
            public float DisplacementSimulator_Dampner;
            public float DisplacementSimulator_StartingSize;
            public ushort Damage;

            public DATAField(UnityBinaryReader r, int dataSize)
            {
                if (dataSize != 102 && dataSize != 86 && dataSize != 62 && dataSize != 42 && dataSize != 2)
                    WindVelocity = 1;
                if (dataSize == 2)
                {
                    Damage = r.ReadLEUInt16();
                    return;
                }
                WindVelocity = r.ReadLESingle();
                WindDirection = r.ReadLESingle();
                WaveAmplitude = r.ReadLESingle();
                WaveFrequency = r.ReadLESingle();
                SunPower = r.ReadLESingle();
                ReflectivityAmount = r.ReadLESingle();
                FresnelAmount = r.ReadLESingle();
                ScrollXSpeed = r.ReadLESingle();
                ScrollYSpeed = r.ReadLESingle();
                FogDistance_NearPlane = r.ReadLESingle();
                if (dataSize == 42)
                {
                    Damage = r.ReadLEUInt16();
                    return;
                }
                FogDistance_FarPlane = r.ReadLESingle();
                ShallowColor = r.ReadT<ColorRef4>(dataSize);
                DeepColor = r.ReadT<ColorRef4>(dataSize);
                ReflectionColor = r.ReadT<ColorRef4>(dataSize);
                TextureBlend = r.ReadByte();
                r.SkipBytes(3); // Unused
                if (dataSize == 62)
                {
                    Damage = r.ReadLEUInt16();
                    return;
                }
                RainSimulator_Force = r.ReadLESingle();
                RainSimulator_Velocity = r.ReadLESingle();
                RainSimulator_Falloff = r.ReadLESingle();
                RainSimulator_Dampner = r.ReadLESingle();
                RainSimulator_StartingSize = r.ReadLESingle();
                DisplacementSimulator_Force = r.ReadLESingle();
                if (dataSize == 86)
                {
                    //DisplacementSimulator_Velocity = DisplacementSimulator_Falloff = DisplacementSimulator_Dampner = DisplacementSimulator_StartingSize = 0F;
                    Damage = r.ReadLEUInt16();
                    return;
                }
                DisplacementSimulator_Velocity = r.ReadLESingle();
                DisplacementSimulator_Falloff = r.ReadLESingle();
                DisplacementSimulator_Dampner = r.ReadLESingle();
                DisplacementSimulator_StartingSize = r.ReadLESingle();
                Damage = r.ReadLEUInt16();
            }
        }

        public struct GNAMField
        {
            public FormId<WATRRecord> Daytime;
            public FormId<WATRRecord> Nighttime;
            public FormId<WATRRecord> Underwater;

            public GNAMField(UnityBinaryReader r, int dataSize)
            {
                Daytime = new FormId<WATRRecord>(r.ReadLEUInt32());
                Nighttime = new FormId<WATRRecord>(r.ReadLEUInt32());
                Underwater = new FormId<WATRRecord>(r.ReadLEUInt32());
            }
        }

        public override string ToString() => $"WATR: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public STRVField TNAM; // Texture
        public BYTEField ANAM; // Opacity
        public BYTEField FNAM; // Flags
        public STRVField MNAM; // Material ID
        public FMIDField<SOUNRecord> SNAM; // Sound
        public DATAField DATA; // DATA
        public GNAMField GNAM; // GNAM

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = r.ReadSTRV(dataSize); return true;
                case "TNAM": TNAM = r.ReadSTRV(dataSize); return true;
                case "ANAM": ANAM = r.ReadT<BYTEField>(dataSize); return true;
                case "FNAM": FNAM = r.ReadT<BYTEField>(dataSize); return true;
                case "MNAM": MNAM = r.ReadSTRV(dataSize); return true;
                case "SNAM": SNAM = new FMIDField<SOUNRecord>(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize); return true;
                case "GNAM": GNAM = new GNAMField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}